### ISSUE TITLE
Master Issue for Updating Color/Attribute System

### ISSUE THREAD

**Issue by tig** [gui-cs/Terminal.Gui#457]
This is the master Issue for revamping the Color system in v2. Dependencies:

- #4086
- #4062
- #4057 
- #4058
- #2381
- #2797
- #2798
- #2799 
- #219 
- #2801
- #2800

See also: 

- #415

**Comment by BDisp** [comment]
@tig the link is the same as this.

**Comment by tig** [comment]
The fix for #666, PR #2612 is providing groundwork for this. Anyone choosing to start on this PR should base it on that PR if you do it before merged in to v2_develop.

**Comment by adstep** [comment]
Hey @tig, is #2612 still the right branch to start from if I'm playing around with porting over the changes from #1628?

**Comment by tznind** [comment]
> Hey @tig, is #2612 still the right branch to start from if I'm playing around with porting over the changes from #1628?

That's correct, it is the most recent console driver refactoring branch.  Please can you consider taking/adapting this https://github.com/veeman/gui.cs/pull/2 (tznind:true-color) PR which I opened into https://github.com/gui-cs/Terminal.Gui/pull/1628 it adds the following:

- UICatalog scenario for opening images
- Reverts the new TrueColorAttribute change to instead make `Attribute` implicitly support/true at the same time with fall back to basic colors when not supported by driver/environment.

I would really love true color support for so many cool things.  Especially for colored icons in FileDialog, gradient borders etc.  Let me know if you want to collaborate on this.

**Comment by adstep** [comment]
Sure I can give it a try this week. Seems straightforward enough given the pre-existing examples.

Is this the right place to loop back if I have any troubles?

**Comment by tznind** [comment]
> Sure I can give it a try this week. Seems straightforward enough given the pre-existing examples.

Awesome, this is really exciting!
 
> Is this the right place to loop back if I have any troubles?

Sure, post any initial/logistics issues here.  When there is a PR (even a draft one) then we can talk about more specifics there.  

I guess since  #2612 is +10,135/−9,977 vs `v2_develop` it makes most sense to branch off `tig:v2_fixes_666_console_driver_dupes` and probably target tigs branch with a draft PR with the updated true color code.  Otherwise the diff will be lost in all the other changes (a lot are just layout) and hard to review.  

We can always retarget once `tig:v2_fixes_666_console_driver_dupes` is merged if @tig wants to merge them in seperately.

**Comment by tig** [comment]
It will likely be a week or so before I can spend time on my console dupe PR, so I concur you should party on it and submit PRs to it for the color stuff. 

Really excited to see someone picking this up! Thank you!

**Comment by adstep** [comment]
Got it running. See https://github.com/tig/Terminal.Gui/pull/15.

**Comment by BDisp** [comment]
> Got it running. See [tig#15](https://github.com/tig/Terminal.Gui/pull/15).

This is a great work. Thanks for providing this. Does it also works on another drivers (`CursesDriver` and `NetDriver`)?

**Comment by tznind** [comment]
Its ok if it only works for console driver to start with.

We can always add support in the other drivers after.  As long as the fallback mechanism is working and the true color class design is sound.

The original work was done by @veeman so it might be we don't have the expertise between us to do the other drivers yet.

**Comment by adstep** [comment]
> Does it also works on another drivers (`CursesDriver` and `NetDriver`)?

It doesn't work on the `CursesDriver` or `NetDriver`. 

The PR just enables WindowsDriver to use VT sequences to describe the requested console color. This should work consistently because there's a clear mechanism to detect whether TrueColor support is possible by checking the version of Windows.

AFAIK Curses uses a mechanism where, if you aren't using one of the 16 default Console colors, you have to register the color you want to use and get returned an identifier to pass along for future calls. If you want to use VT sequences, when using Curses, you'd need to figure out some other means of detecting whether they'd be supported.

**Comment by tznind** [comment]
> If you want to use VT sequences, when using Curses, you'd need to figure out some other means of detecting whether they'd be supported.

I think this should be handled with config files in the same way we support Nerd Fonts (#2613).

So any end user could add to their `.tui` folder config:

```json
"VTTrueColor.Enable": true
```

Or if the vast majority of modern consoles support this we could default it to true and let end users opt out with a setting of false in config.

**Comment by BDisp** [comment]
@adstep I don't know if you already seen my explanation why some colors has the value -1, in https://github.com/tig/Terminal.Gui/pull/15#discussion_r1243574508.

**Comment by BDisp** [comment]
Great stuff here:
https://gist.github.com/fnky/458719343aabd01cfb17a3a4f7296797?permalink_comment_id=4619910#gistcomment-4619910

**Comment by tznind** [comment]
@BDisp is it still possible to output specific escape sequences in CursesDriver?

I know @tig mentioned reducing/eliminating reliance on ncurses methods in https://github.com/gui-cs/Terminal.Gui/pull/2612#issuecomment-1565264597 in favour of direct use of escape codes (he references https://github.com/gui-cs/Terminal.Gui/discussions/1107)

I found and tested some of the console escape codes in the following video on my linux box and found they work.  Also sounds like most modern linux terminals are going to support at least 256 if not true color escapes:
https://www.youtube.com/watch?v=RKPsd4tD9dQ

See this test script that shows how to output escape sequences for 256 colors
https://github.com/BrodieRobertson/scripts/blob/ae0077d38777151e07824410fade622eb93f0ca0/color256#L4

If the escape codes approach taken in https://github.com/tig/Terminal.Gui/pull/15 would still work in linux then I think we should just be practical about things.  Specifically:

- Assume that 256 colors are supported
- Provide a way to switch to 16 or true color  (via `ConfigurationManager`)
- Ideally centralize the logic in base `ConsoleDriver`
- Ignore the fact that user can remap 256 / 16 color to custom colors (if a user uses custom color schemes we can assume they are happy for that to be applied universally).

Sorry that I am a little late to the party on this as I haven't gotten so deep into all the various schemes etc

**Comment by BDisp** [comment]
@tznind I really don't know, because the `CursesDriver` call `Refresh` to write to the screen. In `NetDriver` we control the writing to the screen and we send escape sequences to a `StringBuilder` and work well. In `CursesDriver` maybe is possible to write escape sequences strings to the buffer and perhaps the `Refresh` can send well them to the screen. But I'm not certain.

**Comment by tig** [comment]
Based on my research I believe @tznind is right.

I am going to resist trying to do this in my huge pr, so I can get it merged asap, but I'm now convinced we can actually replace BOTH CursesDriver and WindowsDriver with a new "AnsiDriver".

**Comment by tig** [comment]
Oh, and if we do end up keeping support for 16 bit color, it will NOT be the default. 

Based on what I know now, I want True Color to be default with "as smart as possible" fallback to 256 and/or 16. 

I worry about performance emitting stuff ourselves. It's easy to make it work, but my intuition is it's also easy to emit way too many sequences potentially causing perf problems.

**Comment by tig** [comment]
#2612 now passes all unit tests - The true color support is fragile though. 

I'm going to refactor a bunch of it in line with what I said above in next few days. Hopefully I can get it to a point where it's architecturally correct, if not completely working, in the next few days. I then have a few more issues in #2612 to address before it can be merged.

**Comment by tig** [comment]
Proposal:

- Change `Color` to be a class. Merge `TrueColor` into it, so that there is a single class that represents a color. Default behavior will be 24-bit/TrueColor. Build in "16 color/256 color fallback"; including letting devs do `Color.Red` etc... for well known color names.
- Get rid of `TrueColor` as a separate class, and simplify `Attribute` to just hold two `Color` members as it did previously
- 'Remove the concept of `Initialized` and `HasValidColors` from `Attribute` given a 24-bit color is always valid. 
- Start with WindowsDriver (which already mostly works). Then get CursesDriver working. Then NetDriver.
- Keep all three existing drivers for now. Make building a stronger "ANSI" back-end another Issue.

Once the above is done, I can merge to v2_develop.

We can then open PRs to fix/adapt other things:

- Update ColorPicker
- Combine/Simpifly the color related scenarios 

@BDisp, @tznind, @adstep - let me know if you have more thoughts, questions, or suggestions on all this!

**Comment by BDisp** [comment]
Did the PR #2729 resolved true color on Linux?

**Comment by tig** [comment]
> Did the PR #2729 resolved true color on Linux?

I don't know. I misunderstood what was going on, and didn't realize it was  different set of work than what @adstep did. So I just closed #2729 without looking at it.

Not sure how to handle this now that you've pointed it out.

**Comment by tig** [comment]
Are you sure it does? I peeked at the code briefly and saw no evidence of related changes to CursesDriver. 

I'll look again later.

**Comment by tig** [comment]
Found this:

https://github.com/WilliamRagstad/ANSIConsole

Might be useful.

**Comment by tig** [comment]
Really good stuff by @YourRobotOverlord here: https://github.com/gui-cs/Terminal.Gui/discussions/4056

The code he's written and the discussion points to a few examples of why the existing `ColorSchemes` design is flawed. It also illustrates how devs want even more control and flexibility than is currently provided.

> When building this, I remember thinking that I wished that all views checked ColorSchemes for a scheme to use the same way that Dialog or Menu, for example, do.
>
> I.e.
> If I set the "Dialog" scheme `Terminal.Gui.Colors.ColorSchemes["Dialog"] = new ColorScheme();`, it is automatically used for all new dialogs.
>
> But if I set the "ListónicoView" scheme, it is not used by default, I must use `listView.ColorScheme = Terminal.Gui.Colors.ColorSchemes["ListView"]`

### ColorSchemes ["xyz"] Does Not Map to View Types, Is Not Type Safe, and Doesn't Scale

Even today, with just 4 default entries, using a string to identify a color scheme is not type safe. It also doesn't really make sense. There's no view of type "Base" for example. "Menus" isn't "Menu", "MenuBar", and "MenuItem".

@YourRobotOverlord illustrates a desire to have settable PER-VIEW ColorSchemes...

### Should Every View Have a Settable Default ColorScheme?

I'm not convinced. I think the concept that Miguel first introduced when he invented ColorSchemes was the right idea (just not implemented great). The idea was/is to define ColorSchemes using higher-level constructs:

- The main app frame - (Today: "Toplevel" or "Base")
- Normal Dialogs and other modal pop-ups - (Today: "Dialog")
- Menus and StatusBars - (Today: "Menu")
- Error situations - (Today: "Error")

I think users want consistency across an app, not dozens of different looks depending on what part of the app they're in. You don't see that much flexibility in any GUI, and TUI apps tend to be even more simplistic.

In all of @YourRobotOverlord's screenshots (I'd love to see more!), his app is completely consistent in this way: He's got an overall color scheme that is applied consistently everywhere. He's built his app such that there's fine-grained control, but from what I see, he's not actually USING that control.

The current model drives consistency across an app via inheritance of ColorSchemes up the SubView hierarchy. If a View's `ColorScheme` is null, the `SuperView`'s scheme will be used. I think this is correct.

**If we do implement a "Per-View Default ColorScheme", it must, by default, not break this.**

### Are `Colors` and `ColorScheme` the Correct Names?

No. Absolutely not. They contain `Attributes` which TODAY only indicate a color. But very shortly will support:

- Bright/Bold
- Dim
- Underline
- Blink
- Reverse
- Hidden
- Strikethrough

In addition, the `Colors` dictionary is static, which is bad juju (I've recently tried to make it thread-safe, but I still see random unit test failures).

Proposal:

- Remove `Colors` and just have devs use `ConfigurationManager.Themes`.
- Rename `ColorScheme` to `Scheme` (or `Attributes`?)

### What Attributes Should `ColorScheme` Support?

Today we have (descriptions pulled from the actual API docs):

- **Normal**: For text when the view is not focused, hot, or disabled.
- **HotNormal**: For text in a non-focused view that indicates a HotKey.
- **Focus**: For text in a focused view.
- **HotFocus**: For text in a focused view that indicates a HotKey.
- **Disabled**: For text when the view is disabled.

These work ok, but we've heard loud and clear from devs that they want more. Here's my current proposal for a short-term, simple fix that DOES NOT help with #2381, with revised/clarified descriptions:

- **Normal**: The default attributes if no other attribute is used.
- **HotNormal**: The default attributes if no other attribute is used and the element indicates a HotKey will activate it.
- **Focus**: For elements that indicate the View has focus.
- **HotFocus**: For elements that indicate the View has focus and the element indicates a HotKey will accept it.
- **Active**: For elements within a View that are active (selected). E.g., the selected item in a ListView or FlagSelector.
- **HotActive**: For elements within a View that are active (selected) and the element indicates a HotKey will accept it.
- **Disabled**: For elements when the view is disabled.
- **ReadOnly**: For elements that are normally editable but are currently read-only. E.g., for text within a TextView.

### Aligning with Terminal Settings

To address #2381 (honor the user's terminal settings), we need to map Terminal.Gui’s `Scheme` attributes to user-defined terminal colors, particularly for output styling. PowerShell’s `$PSStyle` provides a robust model, with `$PSStyle.Formatting` defining output colors and `$PSStyle.PSReadLine` offering some relevant input styles. After analyzing potential `$PSStyle.PSReadLine` mappings (`Default → Normal`, `Emphasis → HotNormal`, `Comment → ReadOnly`, `Selection → Active`, `Error → Error`), I propose the following refined mappings, focusing on output styling and minimizing syntax-specific tokens:

- **Normal**: Map to `$PSStyle.Foreground` or `$Host.UI.RawUI.ForegroundColor` for the terminal’s default text color. This is more general than `$PSStyle.PSReadLine.Default`, ensuring broad applicability to output in views like `Label` or `ListView`.
- **Active**: Map to `$PSStyle.PSReadLine.Selection` for selected elements (e.g., highlighted `ListView` items, `RadioGroup` options), aligning with PowerShell’s selection style.
- **Error**: Map to `$PSStyle.Formatting.Error` for error messages or states (e.g., invalid `TextField` input), reflecting PowerShell’s output error color.
- **Warning** (new): Map to `$PSStyle.Formatting.Warning` for warning messages, supporting nuanced output styling.
- **Header** (new): Map to `$PSStyle.Formatting.TableHeader` for table/list headers (e.g., `TableView`, `ListView`).

**Why Skip Some PSReadLine Tokens?**
- `$PSStyle.PSReadLine.Emphasis → HotNormal`: `Emphasis` (for highlighted text like search matches) doesn’t align with `HotNormal`’s HotKey role. A new `Highlighted` attribute could be added later if needed.
- `$PSStyle.PSReadLine.Comment → ReadOnly`: `Comment` is syntax-specific (for code comments), not a UI state like `ReadOnly`. This is better suited for syntax highlighting improvements.
- `$PSStyle.PSReadLine.Error`: Less relevant than `$PSStyle.Formatting.Error`, which is designed for output errors.

These mappings ensure Terminal.Gui reflects PowerShell’s user-defined colors for output, addressing #2381. For non-PowerShell terminals, we can fall back to `$TERM` capabilities (e.g., 16, 256, or 24-bit RGB) or ANSI defaults. Terminal.Gui must support ANSI styles (Bold, Italic, Reverse, etc.) and enforce contrast (e.g., WCAG 2.1’s 4.5:1 ratio) for accessibility.

### Next Steps for #2381
- **Terminal Queries**: Develop a mechanism to query `$PSStyle` (e.g., via PowerShell script execution) or `$TERM` for color capabilities. This is TBD but critical for dynamic color adoption.
- **Syntax Highlighting**: Separately, enhance the `SyntaxHighlighting` sample scenario to leverage `$PSStyle.PSReadLine` tokens like `Comment`, `Keyword`, and `String` for PowerShell script rendering in `TextView` or `ConsoleView`.

**Comment by BDisp** [comment]
`Scheme` is more a generic term and can be apply to a bunch of things. `ColorsScheme` or `AttributesScheme` may sound great.

**Comment by YourRobotOverlord** [comment]
>@YourRobotOverlord illustrates a desire to have settable PER-VIEW ColorSchemes...

Here's an example of where per-view ColorSchemes are necessary:
***Using a ListView-specific ColorScheme***
Notice as focus changes between a TextField and a ListView. The Input color scheme (bright green **foreground** here) is used consistently throughout my app to indicate that an item is editable. The highlighted input color scheme (bright green **background** here) indicates that pressing any key will likely replace the highlighted item - in this case, selected text or the selected item in a ListView.

![Image](https://github.com/user-attachments/assets/7a38fa8f-2766-416f-b34c-cc5293e62146)

***ListView inherits ColorScheme from parent Dialog***
In this scenario, I am unable to keep this visual language consistent because attributes such as `Normal` and `Focus` have a different contextual meaning to a Dialog than to a ListView.

![Image](https://github.com/user-attachments/assets/2526ea39-c111-45cb-ae78-020b7b8cb816)

>Should Every View Have a Settable Default ColorScheme?
>
>I'm not convinced.

I would agree that every View having a settable _default_ ColorScheme may be overkill.
I currently set the ListView ColorScheme in a subclassed Dialog and this seems like a perfectly reasonable way to set a default scheme only where it is needed.

> The current model drives consistency across an app via inheritance of ColorSchemes up the SubView hierarchy. If a View's ColorScheme is null, the SuperView's scheme will be used. I think this is correct.
> 
> If we do implement a "Per-View Default ColorScheme", it must, by default, not break this.

I completely agree, this behavior is very friendly and intuitive.

**Comment by tig** [comment]
Dupe.
